### PR TITLE
Themes

### PR DESCRIPTION
**Pull request by luiseeo**
TLV build with changes in \\tsi.lan\files\Home\AMER\lenciso\Shared\tlv_qt\tlv.win.themes\

- Added a drop down menu in the Options dialog to select the theme
- There are 4 themes right now: Native, Dark Fusion, Solarized Light and Solarized Dark
- A bunch of images where edited or added to that both Dark and Light themes look all right.
- When highlighting entries, a contrasting color for the font is chosen. This is to avoid cases, for example, where the highlight color is bright and the theme font color is also bright making it unreadable.
- Removed the "Gray background" option
- Tested in Windows and Mac (needed to fix some specific details in both)

### CODE REVIEW

File: src/themeutils.cpp [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    double dark = std::min(luminance1, luminance2);
+    return (light + 0.05) / (dark + 0.05);
+}
+
+namespace ThemeUtils {
+    QFileSelector m_fileSelector;
+    void SetLightIconSet()
+    {
+        QStringList extraSelectors;
+        //extraSelectors << "lighttheme";
```
Comment (terencesoh) [inline src/themeutils.cpp]: Can remove

File: src/treemodel.cpp [hunk 2]
```diff
@@ -46,14 +47,25 @@ QVariant TreeModel::data(const QModelIndex &index, int role) const
...
-            break;
+            }*/
+            QColor highlightColor(ItemHighlightColor(index));
+            if (highlightColor == Qt::transparent)
+            {
+                // do nothing if there's no highlight color
+                break;
+            }
+            // Pick a White or Black foreground, depending on which one gives better contrast
+            double whiteContrast = ThemeUtils::ContrastRatio(QColor(Qt::white), highlightColor);
```
Comment (terencesoh) [inline src/treemodel.cpp]: Potential to cache it like the bg color. Does not look bad though.

File: src/optionsdlg.cpp [hunk 3]
```diff
@@ -159,6 +153,12 @@ void OptionsDlg::on_serviceEnable_clicked()
     ui->serviceURLEdit->setEnabled(true);
 }
 
+void OptionsDlg::on_themeComboBox_currentTextChanged(const QString &themeName)
+{
+    qDebug() << "Theme: " << themeName;
+    ThemeUtils::SwitchTheme(themeName, parent);
```
Comment (luiseeo) [inline src/optionsdlg.cpp]: Good observation. I will try to revert if the users clicks Cancel

File: src/optionsdlg.cpp [hunk 4]
```diff
@@ -3,15 +3,18 @@
...
 #include <QBitArray>
 #include <QDebug>
 #include <QFileDialog>
 #include <QInputDialog>
+#include <QPalette>
 #include <QSettings>
 
 OptionsDlg::OptionsDlg(QWidget *parent) :
     QDialog(parent),
+    parent(parent),
```
Comment (luiseeo) [inline src/optionsdlg.cpp]: Good point

File: src/resources.qrc [hunk 5]
```diff
@@ -28,4 +28,38 @@
...
+        <file alias="ctx-up.png">../resources/images/darktheme/ctx-up.png</file>
+        <file alias="ctx-hide.png">../resources/images/darktheme/ctx-hide.png</file>
+        <file alias="ctx-diff.png">../resources/images/darktheme/ctx-diff.png</file>
+        <file alias="ctx-book.png">../resources/images/darktheme/ctx-book.png</file>
+        <file alias="ctx-copy.png">../resources/images/darktheme/ctx-copy.png</file>
+        <file alias="ctx-open-file.png">../resources/images/darktheme/ctx-open-file.png</file>
+        <file alias="value-next.png">../resources/images/darktheme/value-next.png</file>
+        <file alias="value-previous.png">../resources/images/darktheme/value-previous.png</file>
+    </qresource>
+    <qresource prefix="/+darktheme">
```
Comment (luiseeo) [inline src/resources.qrc]: hmmm, weird... there are 2 darktheme sections. I probably did a bad merge at some point

File: src/theme.cpp [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+#include <QPalette>
+#include <QStyle>
+#include <QStyleFactory>
+#include <QWidget>
+
+std::unique_ptr<QPalette> Theme::GetNativePalette(QWidget* widget)
+{
+    return std::unique_ptr<QPalette>(
+        new QPalette(widget->style()->standardPalette())
+    );
```
Comment (luiseeo) [inline src/theme.cpp]: yep, for some reason I used make_unique anywhere but here

File: src/theme.h [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    static std::unique_ptr<QPalette> GetSolarizedLightPalette();
+    static std::unique_ptr<QPalette> GetSolarizedDarkPalette();
+    static QString GetDefaultStyle();
+    static QString sm_defaultStyle;
+
+    bool m_isFusionStyle;
+    bool m_isDark;
+    std::unique_ptr<QPalette> m_palette;
+public:
+    Theme(bool isFusionStyle, bool isDark, std::unique_ptr<QPalette> palette);
```
Comment (luiseeo) [inline src/theme.h]: I tried moving it to private, but then make_unique<Theme> starts complaining it can't access the constructor

**Comment by luiseeo** [comment]
Pushed a commit addressing Terence's comments

**Pull request merged**